### PR TITLE
Finetuning doc + freezing parameters option

### PR DESCRIPTION
**Pull request by b-flo**
Hi,

This PR add freezing parameters option + documentation for finetuning options (transfer learning + freezing). Feel free to tell me or modify docs if it lacks clarity!

### CODE REVIEW

File: doc/tutorial.md [hunk 1]
```diff
@@ -238,7 +238,21 @@ Basically, this option makes training iteration faster than `--batch-count seq`.
...
+
+1.2. Additionally, a list of modules (separated by a comma) can be specified to control the modules to transfer using `--enc-init-mods` and `--dec-init-mods` options.
+It should be noted the user doesn't need to specify each module individually, only a partial matching (beginning of the string) is needed.
+
+Example 1: `--enc-init-mods='enc.'` means all encoder modules should be transfered.
+Example 2: `--enc-init-mods='enc.embed.,encoder.0.'` means embedding layer and first layer should be transfered.
+
+2. Freezing option can be used through `--freeze-mods`. Similarly to `--(enc|dec)-init-mods`, the option take a list of modules (separated by a comma). The behaviour being the same (partial matching).
+
+Example: `--freeze-mods='enc.embed'` means all encoder modules should be frozen.
```
Comment (sw005320) [inline doc/tutorial.md:255]: Maybe you can add an example of multiple arguments separated by a comma?

File: doc/tutorial.md [hunk 2]
```diff
@@ -238,7 +238,21 @@ Basically, this option makes training iteration faster than `--batch-count seq`.
 
     This creates the minibatch that has the maximum number of input, output and input+output frames under 800, 100 and 900, respectively. You can set one of `--batch-frames-xxx` partially. Like `--batch-bins`, this option makes training iteration faster than `--batch-count seq`. If you already has the best `--batch-seqs x` config, try `--batch-frames-in $((x * (mean(ilen) * idim)) --batch-frames-out $((x * mean(olen) * odim))`.
 
+### How to use finetuning
 
+ESPnet currently supports two finetuning operations: transfer learning (1.x) and freezing (2.).
+
+1.1. Transfer learning option is split between encoder initialization (`--enc-init`) and decoder initialization (`--dec-init`). However, the same model can be specified for both options. Each one taking a snapshot or model path.
```
Comment (sw005320) [inline doc/tutorial.md:245]: Can you clarify this sentence a bit more (with some examples?)? 
"Each one taking a snapshot or model path."

File: doc/tutorial.md (hunk 1 above)
Comment (b-flo) [reply doc/tutorial.md]: Done! There was some mistakes too, sorry

File: doc/tutorial.md (hunk 2 above)
Comment (b-flo) [reply doc/tutorial.md]: Is that okay like this?

**Comment by b-flo** [comment]
@kan-bayashi I added freezing to noam for TTS. I also modified asr_mix, mt and st according to the new options.

**Comment by b-flo** [comment]
Hum... I don't know what happened here? Could someone highlight me on the commits added to this PR? :O

File: doc/tutorial.md (hunk 2 above)
Comment (sw005320) [reply doc/tutorial.md]: OK for me!

**Comment by kamo-naoyuki** [comment]
It's better to reset the last merge commit and merge develop branch.

(BTW, I'll plan to remove the develop branch and unify to master branch at the next release because it's confusing for us)

**Comment by kamo-naoyuki** [comment]
CI is failing.

**Comment by b-flo** [comment]
In fact, I think the revert suffice as there should not be any conflict. However, I have the following error in CI and I'm not too familiar with black:

```
would reformat /home/runner/work/espnet/espnet/espnet/asr/pytorch_backend/asr_mix.py
would reformat /home/runner/work/espnet/espnet/espnet/mt/pytorch_backend/mt.py
would reformat /home/runner/work/espnet/espnet/espnet/st/pytorch_backend/st.py
Oh no! 💥 💔 💥
3 files would be reformatted, 524 files would be left unchanged.
Please apply:
    $ black espnet espnet2 test utils setup.py egs*/*/*/local egs2/TEMPLATE/asr1/pyscripts
##[error]Process completed with exit code 1.
```

**Comment by kamo-naoyuki** [comment]
Read the message and follow.

**Comment by b-flo** [comment]
Oh my, sorry. I didn't have black and I didn't think I had to do it myself... :disappointed_relieved:

**Comment by b-flo** [comment]
@kamo-naoyuki I'm trying to figure out why CI fails for (test-*) here: 

```
# (in test file test_utils/test_spm.bats, line 25)
#   `diff ${bpemodel}.vocab test/tedlium2.vocab' failed
```

The test pass successfully on my side however there are some issues for theses tests: mismatched order + probs, and user_defined_symbols not included in generated vocab.

**Comment by kamo-naoyuki** [comment]
I'm not sure why circle-ci is failed but github-runner is successed.
The error comes from sentencepiece's update and I already fixed it, so this is no problem. I'll merge this PR.

**Comment by Cescfangs** [comment]
Are these finetuning staff available for espnet2?

**Comment by b-flo** [comment]
Hi,

A quick glance into espnet2 code tells me that it's not available yet. @kamo-naoyuki, do you confirm?
I'm working on porting transducer to espnet2, I can add them at the same time I guess. It won't be available before next week though.

**Comment by kamo-naoyuki** [comment]
@Cescfangs  Yes, we have. 

`--pretrain_path` and `--pretrain_key`

https://github.com/espnet/espnet/blob/3ea59a0050e8a6a40138ac2365c258825b02f9cd/espnet2/tasks/abs_task.py#L1076-L1090

**Comment by Cescfangs** [comment]
> @Cescfangs  Yes, we have. 
> 
> `--pretrain_path` and `--pretrain_key`
> 
> https://github.com/espnet/espnet/blob/3ea59a0050e8a6a40138ac2365c258825b02f9cd/espnet2/tasks/abs_task.py#L1076-L1090
> 

thanks

**Pull request merged**